content/blog/autopath.md

```diff
@@ -0,0 +1,153 @@
+(limited to 3 more). So, search path can get quite deep, up to 6 domains long.
+
+The ClusterFirst DNS Policy also sets ndots to 5. This means that virtually every query a pod makes
+is run through the search path. A query name would need to contain 5 dots to “skip” the search path.
+
+So, why 5 ndots? The reason for this high ndots setting is the due to the potentially high number of
+dots in a local service’s name.  For example, consider an SRV query for
+`_http._tcp.service.namespace.svc`.  This query can be resolved using the 3rd domain in the search
+path, trying `_http._tcp.service.namespace.svc.cluster.local`.  If ndots was set to something less
+than 5, then the query would be the absolute name `_http._tcp.service.namespace.svc.` which would
+not produce an answer.
+
+## High ndots
+
+The combination of long search path and high ndots means that almost every query made is eligible to
+be searched on a long list of domains before finding an answer.  This has the most impact with
+queries for external resources.  Any search for an external host with less than 5 dots will be run
+through the whole search list before trying the absolute name.
+
+This problem can manifest itself as scaling problems, such as high latency of DNS responses, high
+load on the DNS server, and network congestion.
+
+You can experiment with *autopath* yourself (even without Kubernetes), consider the following
+Corefile:
+
+~~~ corfile
+. {
+    proxy . 8.8.8.8
+    autopath . resolv.conf
+    log
+}
+~~~
+
+And this file `resolv.conf` with a search path.
+
+~~~ txt
+search svc.cluster.local cluster.local local
+~~~
+
+For some extra logging in CoreDNS I'm using [this
+patch](https://gist.github.com/miekg/7c5d4176c82a49e7b9a8d2d18249e421). Starting CoreDNS and asking
+for `google.com.svc.cluster.local` we get the following answer:
+
+~~~ txt
+% dig @localhost A google.com.svc.cluster.local +noall +answer                                                          ~
+
+google.com.svc.cluster.local. 299 IN	CNAME	google.com.
+google.com.		299	IN	A	216.58.206.110
+~~~
+
+*autopath* performed the search path queries on the server and returned the answer with a CNAME
+connecting the final answer to the original question asked.
+
+In the logs below note the three unsuccessful queries before the final successful result.
+
+~~~ txt
+2017/10/15 09:36:00 [INFO] Querying google.com.svc.cluster.local.
+2017/10/15 09:36:00 [INFO] NXDOMAIN for google.com.svc.cluster.local., continuing search
+2017/10/15 09:36:00 [INFO] Querying google.com.cluster.local.
+2017/10/15 09:36:00 [INFO] NXDOMAIN for google.com.cluster.local., continuing search
+2017/10/15 09:36:00 [INFO] Querying google.com.local.
+2017/10/15 09:36:00 [INFO] NXDOMAIN for google.com.local., continuing search
+2017/10/15 09:36:00 [INFO] Querying google.com.
+::1 - [15/Oct/2017:09:36:00 +0100] "A IN google.com.svc.cluster.local. udp 58 false 4096" NOERROR qr,rd,ra 98 94.736986ms
+~~~
```

Comment:
I think we need a closing paragraph explaining how auto path helps with the client side latency, etc.